### PR TITLE
Add classifier wrapper for river and capymoa

### LINKED ISSUE #563: FR: Add wrapper for river and capymoa classifiers
### Describe the workflow you want to enable

The current implementation of the sklearn wrapper in `river` is not compatible with `scikit-learn` >= 1.6, which we depend on. And capymoa has no inbuilt scikit-learn wrapper. We should offer a wrapper for both river and capymoa such that the classifiers can be integrated more easily.

### Describe your proposed solution

Wrapper classes similar to SklearnWrapper and SkorchWrapper.

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tpham93**
Closes #563

### CODE REVIEW

File: skactiveml/classifier/_wrapper.py [hunk 1]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+        Parameters
+        ----------
+        estimator_class : capymoa.base.MOAClassifier.__class__
+            The `capymoa` classifier class that is used to initialize the
+            `capymoa` classifier.
+        estimator_param_dict : dict, default=None
+            Additional arguments for `capymoa.base.MOAClassifier`. If
+            `estimator_param_dict` is `None`, no additional arguments are
+            added. `schema` is not allowed in this dictionary and will be
+            created internally.
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py:1323]: Is there a reason to use the estimator class as input instead of the object?

File: skactiveml/classifier/_wrapper.py [hunk 2]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+            super().__init__(
+                classes=classes,
+                missing_label=missing_label,
+                cost_matrix=cost_matrix,
+                random_state=random_state,
+            )
+            self.estimator_class = estimator_class
+            self.estimator_param_dict = estimator_param_dict
+
+        def fit(self, X, y):
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py:1374]: Do all CopyMoa estimators comply with the implemented `fit`, `partial_fit`, `predict`, and `predict_proba` signature?

File: skactiveml/classifier/_wrapper.py [hunk 3]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+                    )
+                y_pred = self._le.inverse_transform(y_pred)
+            else:
+                p = self.predict_proba([X[0]])[0]
+                y_pred = self.random_state_.choice(
+                    np.arange(len(self.classes_)), len(X), replace=True, p=p
+                )
+                y_pred = self._le.inverse_transform(y_pred)
+            y_pred = y_pred.astype(self.classes_.dtype)
+            return y_pred
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py]: Can we not use `super.predict()`?

File: skactiveml/classifier/_wrapper.py [hunk 4]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+                Input samples.
+
+            Returns
+            -------
+            P : array-like of shape (n_samples, classes)
+                The class probabilities of the input samples. Classes are
+                ordered according to the attribute `self.classes_`.
+            """
+            import capymoa
+            import capymoa.instance
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py:1431]: Is there a specific reason for this in-method import? Could this lead to import issues during prediction, which should have been already before, e.g., during fitting?

File: skactiveml/classifier/_wrapper.py [hunk 5]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+                P_list = []
+                for x in X:
+                    x_instance = capymoa.instance.Instance(
+                        schema=self.schema_, instance=x
+                    )
+                    P_i = self.estimator_.predict_proba(x_instance)
+                    # if estimator_ fails, it returns None. In this case, we
+                    # use a uniform distribution as fallback
+                    if P_i is None:
+                        P_i = np.ones(n_classes) / n_classes
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py]: Note that in other classifiers, we used the class prior (e.g., normalized label counts) as fallback. You can see also the  code below. When can the prediction fail for a single-instance?

File: skactiveml/classifier/_wrapper.py [hunk 6]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+            import capymoa.instance
+
+            check_is_fitted(self)
+            predict_dict = {"ensure_min_samples": 1, "ensure_min_features": 1}
+            X = check_array(X, **(self.check_X_dict_ | predict_dict))
+            check_n_features(self, X, reset=False)
+            n_classes = len(self.classes_)
+            if self.is_fitted_:
+                P_list = []
+                for x in X:
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py:1440]: Is such an iteration because of the stream nature of CopyMoa necessary?

File: skactiveml/classifier/_wrapper.py [hunk 7]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+            else:
+                return np.tile(
+                    self._label_counts / np.sum(self._label_counts),
+                    [len(X), 1],
+                )
+
+        def _fit(self, fit_function, X, y, sample_weight=None):
+            import capymoa
+            import capymoa.base
+            import capymoa.instance
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py:1480]: See comment regarding in-method imports.

File: skactiveml/classifier/_wrapper.py [hunk 8]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+            ]
+            if hasattr(self, "estimator_"):
+                if fit_function != "partial_fit":
+                    self.estimator_ = self._create_estimator(X)
+            else:
+                self.estimator_ = self._create_estimator(X)
+            if self.estimator_ is None:
+                self.is_fitted_ = False
+                return self
+            # Check whether estimator can deal with cost matrix.
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py]: How is this checked in the code below?

File: skactiveml/classifier/_wrapper.py [hunk 9]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+            X : array-like of shape (n_samples, ...)
+                The feature matrix representing the samples.
+
+            Returns
+            -------
+            estimator: CapyMOAClassifier,
+                The initialized but untrained `CapyMOAClassifier`.
+            """
+            import capymoa
+            import capymoa.stream
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py:1566]: See comment regarding in-method imports.

File: skactiveml/classifier/_wrapper.py [hunk 10]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
                     if self.cost_matrix is None
                     else self.cost_matrix
                 )
+
+
+if successful_capymoa_import:
+
+    class CapyMOAClassifier(SkactivemlClassifier):
+        """CapyMOA Classifier
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py:1307]: Add reference to CapyMoa.

File: skactiveml/classifier/_wrapper.py [hunk 11]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+            self.schema_ = capymoa.stream.Schema.from_custom(
+                features=features, target="label", categories=categories
+            )
+            return self.estimator_class(self.schema_, **estimator_kwargs)
+
+
+if successful_river_import:
+
+    class RiverClassifier(SkactivemlClassifier, MetaEstimatorMixin):
+        """River Classifier
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py:1599]: Add references to River.

File: skactiveml/classifier/_wrapper.py [hunk 12]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+
+        Implementation of a wrapper class for `river` classifiers such that
+        they implement the SkactivemlClassifier interfaces for classifiers.
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py]:
```suggestion
        they implement the `SkactivemlClassifier` interfaces for classifiers.
```

File: skactiveml/classifier/_wrapper.py [hunk 13]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+
+
+if successful_capymoa_import:
+
+    class CapyMOAClassifier(SkactivemlClassifier):
+        """CapyMOA Classifier
+
+        Implementation of a wrapper class for `CapyMOA` classifiers such that
+        missing labels can be handled and the interfaces are compatible with
+        `scikit-learn`. Therefore, samples with missing labels are filtered.
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py]:
```suggestion
        `scikit-activeml`. Therefore, samples with missing labels are filtered.
```

File: skactiveml/classifier/_wrapper.py [hunk 14]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+
+        def fit(self, X, y, sample_weight=None, **fit_kwargs):
+            """Fit the model using `X` as training data and `y` as class
+            labels.
+
+            Parameters
+            ----------
+            X : array-like of shape (n_samples, ...)
+                The feature matrix representing the samples.
+            y : array-like of shape (n_samples,) or (n_samples, n_outputs)
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py:1666]: Should we allow n_outputs > 1. Could this interfere with multi-label?

File: skactiveml/classifier/_wrapper.py [hunk 15]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+        ):
+            super().__init__(
+                classes=classes,
+                missing_label=missing_label,
+                cost_matrix=cost_matrix,
+                random_state=random_state,
+            )
+            self.estimator = estimator
+
+        def fit(self, X, y, sample_weight=None, **fit_kwargs):
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py]: How is a signature mismatch with the underlying estimator handled?

Comment (mherde) [reply skactiveml/classifier/_wrapper.py]: The same comment applies to the other public sklearn API.

File: skactiveml/classifier/_wrapper.py [hunk 16]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+
+            Parameters
+            ----------
+            X : array-like of shape (n_samples, ...)
+                The feature matrix representing the samples.
+            y : array-like of shape (n_samples,) or (n_samples, n_outputs)
+                It contains the class labels of the training samples. Missing
+                labels are represented the attribute `self.missing_label_`. In
+                case of multiple labels per sample (i.e., n_outputs > 1), the
+                samples are duplicated.
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py:1700]: See comment regarding multi-output targets.

File: skactiveml/classifier/_wrapper.py [hunk 17]
```diff
@@ -1282,3 +1300,645 @@ def _initialize_fallbacks(self, P):
...
+            """
+            return self._fit(
+                fit_function="partial_fit",
+                X=X,
+                y=y,
+                sample_weight=sample_weight,
+                **fit_kwargs,
+            )
+
+        def predict(self, X, **predict_kwargs):
```
Comment (mherde) [inline skactiveml/classifier/_wrapper.py]: What about using `super().predict`?

File: tutorials/22_river_classifier.ipynb
Comment (mherde) [inline tutorials/22_river_classifier.ipynb:1]: Could we use a more interesting dataset than make_blobs?

Comment (mherde) [reply tutorials/22_river_classifier.ipynb:1]: This notebook seems also very similar to CopyMOA. Can both be merged into one?

File: tutorials/23_capymoa_classifier.ipynb
Comment (mherde) [inline tutorials/23_capymoa_classifier.ipynb:1]: See comments on river notebook.

File: requirements_opt.txt
Comment (mherde) [inline requirements_opt.txt:1]: Should we split the installations for skorch, river and capymoa? For example, opt installs all optional dependencies, but we also allow installing also single optional dependencies, e.g., skactiveml[skorch], skactiveml[capymoa,river], etc.

**Review by mherde — CHANGES_REQUESTED** [review]

File: requirements_opt.txt
Comment (mherde) [reply requirements_opt.txt:1]: Not necessary.

**Review by mherde — APPROVED** [review]
